Tests/UnitTests/Application/MainLoopTests.cs

```diff
@@ -130,19 +107,21 @@ public void AddIdle_Twice_Returns_False_Called_Twice ()
                                 return true;
                             };
 
-        ml.AddIdle (fnStop);
-        ml.AddIdle (fn1);
-        ml.AddIdle (fn1);
+        var a = ml.TimedEvents.AddTimeout (TimeSpan.Zero, fnStop);
+        var b = ml.TimedEvents.AddTimeout (TimeSpan.Zero, fn1);
         ml.Run ();
-        Assert.True (ml.TimedEvents.RemoveIdle (fnStop));
-        Assert.False (ml.TimedEvents.RemoveIdle (fn1));
-        Assert.False (ml.TimedEvents.RemoveIdle (fn1));
 
-        Assert.Equal (2, functionCalled);
+        Assert.True (ml.TimedEvents.RemoveTimeout(a));
+        Assert.False (ml.TimedEvents.RemoveTimeout (a));
+
+        // Cannot remove b because it returned false i.e. auto removes itself
+        Assert.False (ml.TimedEvents.RemoveTimeout (b));
```

Comment:
I think this is the cause why the `Mainloop_Invoke_Or_AddTimeout_Can_Be_Used_For_Events_Or_Actions` unit test is failing. Maybe the pending idle function isn't executed because it was already removed without call `RemoveTimeout`? If it's the case it's needed to ensure that this timed event come from `Post` method should be run no matter the timed event is elapsed time is expired.

Reply:
Only the `Action a1 = StartWindow;` ran successfully the unit test. The `Action a2 = () => Application.Invoke (StartWindow);` not allowing to be run asynchronous. Probably it isn't being called properly or AddTimeout doesn't allow run non-UIThreads. the `SynchronizationContext.Post` method isn't hit.

```cs
public static IEnumerable<object []> TestAddTimeout
    {
        get
        {
            // Goes fine
            Action a1 = StartWindow;

            yield return new object [] { a1, "Click Me", "Cancel", "Pew Pew", 0, 1, 2, 3, 4 };

            // Also goes fine
            Action a2 = () => Application.Invoke (StartWindow);

            yield return new object [] { a2, "Click Me", "Cancel", "Pew Pew", 0, 1, 2, 3, 4 };
        }
    }
```